src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -9773,11 +9773,27 @@ bool shouldTreatAsModifier()
                 using var _ = this.GetDisposableResetPoint(resetOnDispose: true);
 
                 Debug.Assert(this.CurrentToken.Kind == SyntaxKind.IdentifierToken);
-                this.EatToken();
 
-                return IsDeclarationModifier(this.CurrentToken.Kind) ||
-                    IsAdditionalLocalFunctionModifier(this.CurrentToken.Kind) ||
-                    (ScanType() != ScanTypeFlags.NotType && this.CurrentToken.Kind == SyntaxKind.IdentifierToken);
+                do
+                {
+                    this.EatToken();
+
+                    using var _2 = this.GetDisposableResetPoint(resetOnDispose: true);
+
+                    if (IsDeclarationModifier(this.CurrentToken.Kind) ||
+                        IsAdditionalLocalFunctionModifier(this.CurrentToken.Kind) ||
+                        (ScanType() != ScanTypeFlags.NotType && this.CurrentToken.Kind == SyntaxKind.IdentifierToken))
+                    {
+                        return true;
+                    }
+                }
+                // If the next token is an identifier with some contextual kind, it might be a modifier.
+                // We want to treat current token as modifier if the next token is a modifier, so we need to check that.
+                // This allows to correctly parse things like local functions with several `async` modifiers
+                while (this.CurrentToken.Kind == SyntaxKind.IdentifierToken &&
+                       this.CurrentToken.ContextualKind != SyntaxKind.IdentifierToken);
```

Comment:
Can you just check for async?

Reply:
The original intent is that if there is a new contextual local function modifier added, this code should still work. What is bad with such approach?